servicetalk-concurrent-internal/src/main/java/io/servicetalk/concurrent/internal/DelayedSubscription.java

```diff
@@ -18,84 +18,104 @@
 import io.servicetalk.concurrent.PublisherSource.Subscription;
 
 import java.util.concurrent.atomic.AtomicLongFieldUpdater;
-import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 import javax.annotation.Nullable;
 
-import static io.servicetalk.concurrent.internal.ConcurrentSubscription.wrap;
-import static io.servicetalk.concurrent.internal.EmptySubscription.EMPTY_SUBSCRIPTION;
-import static io.servicetalk.concurrent.internal.SubscriberUtils.isRequestNValid;
+import static io.servicetalk.concurrent.internal.FlowControlUtils.addWithOverflowProtection;
 import static java.lang.Long.MIN_VALUE;
-import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
 
 /**
  * A {@link Subscription} which serves as a placeholder until the "real" {@link Subscription} is available.
  */
 public final class DelayedSubscription implements Subscription {
-    private static final AtomicReferenceFieldUpdater<DelayedSubscription, Subscription> currentUpdater =
-            newUpdater(DelayedSubscription.class, Subscription.class, "current");
+    private static final long SUBSCRIPTION_SETTING = MIN_VALUE;
+    private static final long SUBSCRIPTION_SET = MIN_VALUE + 1;
+    private static final long SUBSCRIPTION_CANCEL_PENDING = MIN_VALUE + 2;
```

Comment:
nit: I like to use `<< 1` etc

Reply:
these defines are the 3 most negative numbers in succession, and not used in bit masking. there is also an optimization since they are sequential and negative most values which uses `<=` to avoid explicit checks of each value when all of them take the same code path.